mpd/client.go

```diff
@@ -16,25 +16,34 @@ import (
 )
 
 // Quote quotes strings in the format understood by MPD.
-// See: http://git.musicpd.org/cgit/master/mpd.git/tree/src/util/Tokenizer.cxx
+// See: https://github.com/MusicPlayerDaemon/MPD/blob/master/src/util/Tokenizer.cxx
+// NB1: double quotes inside double-quoted literals must be escaped by the caller.
+// NB2: this will probably fail should you need to use single quotes outside of double-quoted literal.
 func quote(s string) string {
 	q := make([]byte, 2+2*len(s))
 	i := 0
 	q[i], i = '"', i+1
 	for _, c := range []byte(s) {
-		if c == '"' {
-			q[i], i = '\\', i+1
-			q[i], i = '"', i+1
-		} else {
-			q[i], i = c, i+1
+		// We need to quote single/double quotes and backslash
+		if c == '"' || c == '\\' || c == '\'' {
+			// Prepend the char with a backslash
+			q[i] = '\\'
+			i++
+			// For single quotes the backslash must be doubled
```

Comment:
I'm simply following the original [article on escaping](https://www.musicpd.org/doc/html/protocol.html#escaping-string-values):
> String values are quoted with single or double quotes, and special characters within those values must be escaped with the backslash (\). Keep in mind that the backslash is also the escape character on the protocol level, which means you may need to use double backslash.

> Example expression which matches an artist named foo'bar":

```
(Artist == "foo\'bar\"")
```
> At the protocol level, the command must look like this:
```
find "(Artist == \"foo\\'bar\\\"\")"
```
> The double quotes enclosing the artist name must be escaped because they are inside a double-quoted find parameter. The single quote inside that artist name must be escaped with two backslashes; one to escape the single quote, and another one because the backslash inside the string inside the parameter needs to be escaped as well. The double quote has three confusing backslashes: two to build one backslash, and another one to escape the double quote on the protocol level. 

Now, when it comes to single quotes, when inside double-quoted literals, you either have to escape them externally - which is a hassle - or let the library do it. I chose for the latter simply because I think you'll *only* encounter single quotes inside literals. There's no command or operator with a single quote in it.

Reply:
Our `Find` and `Search` methods were written for the old syntax:

> Prior to MPD 0.21, the syntax looked like this:
> find TYPE VALUE

which doesn't have this quoting issue. This should work as expected:
```
c.Find("artist", "foo'bar")
```

If you're using the new syntax, quoting it externally is still better than having this kluge in my opinion. Would it solve your problem if we export the quote function, so that you can do:
```
c.Find(fmt.Sprintf("(Artist == %s)", mpd.Quote("foo'bar")))
```

Reply:
How would you search for `foo\bar` ?

Reply:
```go
mpd.Quote(`foo\bar`) => `"foo\\bar"`
mpd.Quote(`(Artist == "foo\\bar")`) => `"(Artist == \"foo\\\\bar\")"`
```
I don't see the issue.

Reply:
So if I understood you correctly, you're proposing:
1. Exporting `Quote()`
2. Using the exported `Quote()` for wrapping individual literals
3. Using another `Quote()` call to wrap the entire command before passing it to the protocol

Is that correct? If so, than I foresee the issue in the fact the quoting works differently for **literals** (`"`, `'`, '\' are to be escaped) and **outside literals** (only the outer `"`s and all `\` must be escaped). Or did you mean two separate quoting functions?

Reply:
I'm probably misunderstanding mpd's complex quoting rules. I don't care enough to look deeper into this now. At least we have some tests. We can come back to this later if we have any issues.